{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd7f36b8-8800-498c-bd97-1ec248ab35b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "LAQ 3.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7590a30c-1565-466e-b77c-593efdd457bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Analysis of tumour neoantigen analysis using MuPeXi for WD13\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Load MuPeXi results (Assuming it's a TSV file)\n",
    "df = pd.read_csv(\"C:/Users/dell/Downloads/WD13_Vcf.txt\", sep=\"\\t\", skiprows=5)\n",
    "\n",
    "\n",
    "# Apply filters for tumor neoantigens\n",
    "neoantigens = df[(df[\"Norm_MHCrank_EL\"] > 2) & (df[\"Mut_MHCrank_EL\"] < 2)]\n",
    "\n",
    "# Apply filters for strong tumor neoantigens\n",
    "strong_neoantigens = neoantigens[neoantigens[\"Mut_MHCrank_EL\"] < 0.5]\n",
    "\n",
    "# Count unique neoantigen peptides per tumor\n",
    "neoantigen_counts = neoantigens.groupby(\"Gene_Symbol\").size()\n",
    "\n",
    "# Count strong neoantigens per tumor\n",
    "strong_neoantigen_counts = strong_neoantigens.groupby(\"Gene_Symbol\").size()\n",
    "\n",
    "# Print results\n",
    "print(\"Total Tumor Neoantigens per Tumor Sample of WD13:\")\n",
    "print(neoantigen_counts)\n",
    "\n",
    "print(\"\\nTotal Strong Tumor Neoantigens per Tumor Sample of WD13:\")\n",
    "print(strong_neoantigen_counts)\n",
    "\n",
    "print(f\"Total Tumour Neoantigens of WD13: {len(neoantigens)}\")\n",
    "print(f\"Total Strong Tumour Neoantigens of WD13: {len(strong_neoantigens)}\")\n",
    "\n",
    "# Save results to Excel\n",
    "neoantigens.to_excel(\"Neoantigens_WD13.xlsx\", index=False)\n",
    "strong_neoantigens.to_excel(\"Strong_Neoantigens_WD13.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d243e07-c12c-46aa-9dbc-f1436aab0959",
   "metadata": {},
   "source": [
    "LAQ 3-Part 1:\n",
    "(1) How many tumour neoantigens can you find for this tumour? 366\n",
    "(2) How many strong tumour neoantigens are there for this tumour? 37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "65fc6136-1afb-47fd-abde-6891fde94e12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Tumor Neoantigens per Tumor Sample of WD6:\n",
      "Gene_ID\n",
      "ENSG00000004399    2\n",
      "ENSG00000007174    1\n",
      "ENSG00000008277    1\n",
      "ENSG00000008513    2\n",
      "ENSG00000009307    2\n",
      "                  ..\n",
      "ENSG00000279012    6\n",
      "ENSG00000279219    2\n",
      "ENSG00000279301    2\n",
      "ENSG00000283186    3\n",
      "ENSG00000283652    1\n",
      "Length: 305, dtype: int64\n",
      "\n",
      "Total Strong Tumor Neoantigens per Tumor Sample of WD6:\n",
      "Gene_ID\n",
      "ENSG00000008277    1\n",
      "ENSG00000008513    1\n",
      "ENSG00000071189    2\n",
      "ENSG00000087338    1\n",
      "ENSG00000088179    1\n",
      "ENSG00000090512    1\n",
      "ENSG00000096060    1\n",
      "ENSG00000100281    2\n",
      "ENSG00000100784    1\n",
      "ENSG00000101311    2\n",
      "ENSG00000110427    2\n",
      "ENSG00000113212    1\n",
      "ENSG00000120549    1\n",
      "ENSG00000123485    1\n",
      "ENSG00000127481    1\n",
      "ENSG00000128250    1\n",
      "ENSG00000133124    1\n",
      "ENSG00000136240    1\n",
      "ENSG00000139613    1\n",
      "ENSG00000139629    2\n",
      "ENSG00000140463    1\n",
      "ENSG00000142731    2\n",
      "ENSG00000144476    1\n",
      "ENSG00000149507    1\n",
      "ENSG00000154864    2\n",
      "ENSG00000156232    2\n",
      "ENSG00000157856    1\n",
      "ENSG00000159840    1\n",
      "ENSG00000160145    1\n",
      "ENSG00000160961    1\n",
      "ENSG00000168488    1\n",
      "ENSG00000169876    1\n",
      "ENSG00000169989    1\n",
      "ENSG00000170379    1\n",
      "ENSG00000176571    1\n",
      "ENSG00000179520    1\n",
      "ENSG00000181143    2\n",
      "ENSG00000182185    1\n",
      "ENSG00000186714    1\n",
      "ENSG00000187098    1\n",
      "ENSG00000197496    1\n",
      "ENSG00000221923    1\n",
      "ENSG00000251664    1\n",
      "ENSG00000254466    1\n",
      "ENSG00000254521    4\n",
      "ENSG00000255713    1\n",
      "ENSG00000262209    1\n",
      "ENSG00000263961    4\n",
      "ENSG00000279012    2\n",
      "ENSG00000279301    1\n",
      "dtype: int64\n",
      "Total Tumour Neoantigens of WD6: 570\n",
      "Total Strong Tumour Neoantigens of WD6: 66\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load MuPeXi results (Assuming it's a TSV file)\n",
    "\n",
    "df = pd.read_csv(\"C:/Users/dell/Downloads/WD06.mupexi.txt\", sep=\"\\t\", skiprows=5)\n",
    "\n",
    "\n",
    "# Apply filters for tumor neoantigens\n",
    "neoantigens = df[(df[\"Norm_MHCrank_EL\"] > 2) & (df[\"Mut_MHCrank_EL\"] < 2)]\n",
    "\n",
    "# Apply filters for strong tumor neoantigens\n",
    "strong_neoantigens = neoantigens[neoantigens[\"Mut_MHCrank_EL\"] < 0.5]\n",
    "\n",
    "# Count unique neoantigen peptides per tumor\n",
    "neoantigen_counts = neoantigens.groupby(\"Gene_ID\").size()\n",
    "\n",
    "# Count strong neoantigens per tumor\n",
    "strong_neoantigen_counts = strong_neoantigens.groupby(\"Gene_ID\").size()\n",
    "\n",
    "# Print results\n",
    "print(\"Total Tumor Neoantigens per Tumor Sample of WD6:\")\n",
    "print(neoantigen_counts)\n",
    "\n",
    "print(\"\\nTotal Strong Tumor Neoantigens per Tumor Sample of WD6:\")\n",
    "print(strong_neoantigen_counts)\n",
    "\n",
    "print(f\"Total Tumour Neoantigens of WD6: {len(neoantigens)}\")\n",
    "print(f\"Total Strong Tumour Neoantigens of WD6: {len(strong_neoantigens)}\")\n",
    "\n",
    "# Save results to Excel\n",
    "neoantigens.to_excel(\"Neoantigens_WD6.xlsx\", index=False)\n",
    "strong_neoantigens.to_excel(\"Strong_Neoantigens_WD6.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da0c3864-6c88-4e0b-96f0-d01d121a84bb",
   "metadata": {},
   "source": [
    "LAQ 3-Part 1:\r\n",
    "(1) How many tumour neoantigens can you find for this tumour? 570\r\n",
    "(2) How many strong tumour neoantigens are there for this tumour? 66"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b25239ae-79d1-484d-bd3b-fae46a68879f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Tumor Neoantigens per Tumor Sample of WD14:\n",
      "Gene_Symbol\n",
      "ANKRD16     1\n",
      "AQP4        1\n",
      "ASIC4       1\n",
      "BGLAP       6\n",
      "C1orf168    1\n",
      "CALCOCO2    1\n",
      "CAMSAP3     1\n",
      "CCDC158     1\n",
      "CEP295      1\n",
      "DAG1        1\n",
      "DNAH8       2\n",
      "DPPA5       2\n",
      "FBXL14      1\n",
      "FBXO22      1\n",
      "FLG2        1\n",
      "GTF3C1      1\n",
      "HAUS6       2\n",
      "ITIH6       3\n",
      "KBTBD8      1\n",
      "KCNA2       1\n",
      "LINGO1      1\n",
      "MUC12       1\n",
      "NBAS        3\n",
      "NEDD4       1\n",
      "NUP107      1\n",
      "OR4D9       1\n",
      "PCDH12      1\n",
      "PKD1L1      2\n",
      "PLCB3       2\n",
      "RNF133      1\n",
      "SAMD9L      1\n",
      "SPTBN1      1\n",
      "STK11IP     1\n",
      "SYNE2       1\n",
      "TMEM132C    2\n",
      "TNPO2       2\n",
      "UBAP1       1\n",
      "UBE2Q2      6\n",
      "ZNF714      4\n",
      "ZNF780A     2\n",
      "dtype: int64\n",
      "\n",
      "Total Strong Tumor Neoantigens per Tumor Sample of WD14:\n",
      "Gene_Symbol\n",
      "BGLAP      1\n",
      "CAMSAP3    1\n",
      "DPPA5      1\n",
      "HAUS6      1\n",
      "NBAS       1\n",
      "UBE2Q2     2\n",
      "ZNF780A    1\n",
      "dtype: int64\n",
      "Total Tumour Neoantigens of WD14: 65\n",
      "Total Strong Tumour Neoantigens of WD14: 8\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load MuPeXi results (Assuming it's a TSV file)\n",
    "\n",
    "df = pd.read_csv(\"C:/Users/dell/Downloads/WD14.mupexi.txt\", sep=\"\\t\", skiprows=5)\n",
    "\n",
    "\n",
    "# Apply filters for tumor neoantigens\n",
    "neoantigens = df[(df[\"Norm_MHCrank_EL\"] > 2) & (df[\"Mut_MHCrank_EL\"] < 2)]\n",
    "\n",
    "# Apply filters for strong tumor neoantigens\n",
    "strong_neoantigens = neoantigens[neoantigens[\"Mut_MHCrank_EL\"] < 0.5]\n",
    "\n",
    "# Count unique neoantigen peptides per tumor\n",
    "neoantigen_counts = neoantigens.groupby(\"Gene_Symbol\").size()\n",
    "\n",
    "# Count strong neoantigens per tumor\n",
    "strong_neoantigen_counts = strong_neoantigens.groupby(\"Gene_Symbol\").size()\n",
    "\n",
    "# Print results\n",
    "print(\"Total Tumor Neoantigens per Tumor Sample of WD14:\")\n",
    "print(neoantigen_counts)\n",
    "\n",
    "print(\"\\nTotal Strong Tumor Neoantigens per Tumor Sample of WD14:\")\n",
    "print(strong_neoantigen_counts)\n",
    "\n",
    "print(f\"Total Tumour Neoantigens of WD14: {len(neoantigens)}\")\n",
    "print(f\"Total Strong Tumour Neoantigens of WD14: {len(strong_neoantigens)}\")\n",
    "\n",
    "# Save results to Excel\n",
    "neoantigens.to_excel(\"Neoantigens_WD14.xlsx\", index=False)\n",
    "strong_neoantigens.to_excel(\"Strong_Neoantigens_WD14.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65054ab8-57b4-4404-b4c4-c1a0b315549d",
   "metadata": {},
   "source": [
    "LAQ 3-Part 1:\n",
    "(1) How many tumour neoantigens can you find for this tumour? 65\n",
    "(2) How many strong tumour neoantigens are there for this tumour? 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a7bf183-7d66-4732-8e2c-96b8164c642e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Tumor Neoantigens per Tumor Sample of WD22:\n",
      "Gene_Symbol\n",
      "ABCA1     1\n",
      "ABCA12    1\n",
      "ABCA3     2\n",
      "ABRACL    1\n",
      "ACAD8     1\n",
      "         ..\n",
      "ZFP42     1\n",
      "ZNF214    4\n",
      "ZNF534    1\n",
      "ZNF541    1\n",
      "ZNF692    1\n",
      "Length: 256, dtype: int64\n",
      "\n",
      "Total Strong Tumor Neoantigens per Tumor Sample of WD22:\n",
      "Gene_Symbol\n",
      "ARHGAP15     2\n",
      "ASXL1        1\n",
      "BCO2         1\n",
      "BIN1         1\n",
      "C1S          1\n",
      "CASP12       1\n",
      "CEP295       1\n",
      "CNTLN        1\n",
      "DGKD         1\n",
      "DSEL         1\n",
      "ELAVL4       1\n",
      "EPB42        1\n",
      "EPG5         4\n",
      "FBXO21       1\n",
      "FILIP1       1\n",
      "FLRT2        1\n",
      "GABRB3       1\n",
      "GATA2        2\n",
      "GLTSCR1L     1\n",
      "GPR142       1\n",
      "GRINA        1\n",
      "ICOSLG       4\n",
      "KCNG4        1\n",
      "KIR2DL4      1\n",
      "KLC1         1\n",
      "LAMB4        3\n",
      "MAP7D1       1\n",
      "METTL23      1\n",
      "MROH7        1\n",
      "MYBPC2       2\n",
      "MYO3A        1\n",
      "NRXN1        1\n",
      "OR13C3       1\n",
      "OR1S1        1\n",
      "OR5K3        1\n",
      "PPP3CB       1\n",
      "PSEN2        1\n",
      "PTPRC        1\n",
      "PXDN         1\n",
      "RERG         1\n",
      "RPH3A        1\n",
      "SCML2        1\n",
      "SERPINI2     2\n",
      "SHE          1\n",
      "SLC9C2       1\n",
      "SPATA31E1    1\n",
      "TACC2        1\n",
      "TAS2R38      1\n",
      "TENM4        1\n",
      "TERT         1\n",
      "TKTL2        1\n",
      "TLX3         1\n",
      "TPTE         1\n",
      "TRAJ39       3\n",
      "TSSC1        1\n",
      "ZNF214       1\n",
      "dtype: int64\n",
      "Total Tumour Neoantigens of WD22: 494\n",
      "Total Strong Tumour Neoantigens of WD22: 70\n",
      "       HLA_allele Norm_peptide  Norm_MHCrank_EL  Mut_peptide  Mut_MHCrank_EL  \\\n",
      "0      HLA-A24:02    ......P.E           5.6590    EYSRIVQLF          0.0098   \n",
      "1      HLA-A02:01    .P.......           8.4315    LLTDIVLLL          0.0438   \n",
      "2      HLA-A02:01    .P.......          15.3716    SLIVEEFQV          0.0615   \n",
      "3      HLA-A24:02   ..N.......           5.6694   NNYAGNMLTF          0.0655   \n",
      "4      HLA-A24:02    ........S          14.6989    HFPKWLNGF          0.0887   \n",
      "...           ...          ...              ...          ...             ...   \n",
      "75121  HLA-A02:01  ...........          88.3333  YKGGKIYKVPS         88.3333   \n",
      "75122  HLA-A24:02  ...........          82.5000  YKGGKIYKVPS         82.5000   \n",
      "75123  HLA-A24:02    .........          15.5137    HMDRNPYYG         15.5137   \n",
      "75124  HLA-A02:01  ...........          40.9286  YGGESSSITPL         40.9286   \n",
      "75125  HLA-A24:02    .........          44.5000    RSHASINIG         44.5000   \n",
      "\n",
      "               Gene_ID                                      Transcript_ID  \\\n",
      "0      ENSG00000152223                                    ENST00000282041   \n",
      "1      ENSG00000095777                                    ENST00000265944   \n",
      "2      ENSG00000182326  ENST00000328916,ENST00000360817,ENST0000040394...   \n",
      "3      ENSG00000211850                                    ENST00000390498   \n",
      "4      ENSG00000197580  ENST00000357685,ENST00000361053,ENST0000043802...   \n",
      "...                ...                                                ...   \n",
      "75121  ENSG00000203879                                    ENST00000445564   \n",
      "75122  ENSG00000203879                                    ENST00000445564   \n",
      "75123  ENSG00000203879                                    ENST00000445564   \n",
      "75124  ENSG00000203879                                    ENST00000445564   \n",
      "75125  ENSG00000188227  ENST00000445217,ENST00000542455,ENST0000058714...   \n",
      "\n",
      "      Amino_Acid_Change Allele_Frequency  Mismatches  ...  \\\n",
      "0                   N/X                -           2  ...   \n",
      "1                   P/L                -           1  ...   \n",
      "2                   P/L                -           1  ...   \n",
      "3                   N/Y                -           1  ...   \n",
      "4                   S/F                -           1  ...   \n",
      "...                 ...              ...         ...  ...   \n",
      "75121              S/SX                -           0  ...   \n",
      "75122              S/SX                -           0  ...   \n",
      "75123              S/SX                -           0  ...   \n",
      "75124              S/SX                -           0  ...   \n",
      "75125               C/X                -           0  ...   \n",
      "\n",
      "            Protein_position Mutation_Consequence Gene_Symbol  \\\n",
      "0                        980                    F        EPG5   \n",
      "1                        852                    M       MYO3A   \n",
      "2      94,94,94,94,94,119,76                    M         C1S   \n",
      "3                          4                    M      TRAJ39   \n",
      "4       83,83,49,49,83,49,60                    M        BCO2   \n",
      "...                      ...                  ...         ...   \n",
      "75121                     29                    F        GDI1   \n",
      "75122                     29                    F        GDI1   \n",
      "75123                     29                    F        GDI1   \n",
      "75124                     29                    F        GDI1   \n",
      "75125        285,191,285,285                    F      ZNF793   \n",
      "\n",
      "      Cancer_Driver_Gene Proteome_Peptide_Match Expression_Level  \\\n",
      "0                     No                     No                -   \n",
      "1                     No                     No                -   \n",
      "2                     No                     No                -   \n",
      "3                     No                     No                -   \n",
      "4                     No                     No                -   \n",
      "...                  ...                    ...              ...   \n",
      "75121                 No                    Yes                -   \n",
      "75122                 No                    Yes                -   \n",
      "75123                 No                    Yes                -   \n",
      "75124                 No                    Yes                -   \n",
      "75125                 No                    Yes                -   \n",
      "\n",
      "      Mutant_affinity_score Normal_affinity_score Expression_score  \\\n",
      "0              9.999523e-01          1.133920e-08                1   \n",
      "1              9.999435e-01          1.081870e-14                1   \n",
      "2              9.999383e-01          9.203214e-30                1   \n",
      "3              9.999370e-01          1.076463e-08                1   \n",
      "4              9.999293e-01          2.658821e-28                1   \n",
      "...                     ...                   ...              ...   \n",
      "75121         3.385486e-188         3.385486e-188                1   \n",
      "75122         1.572067e-175         1.572067e-175                1   \n",
      "75123          4.522445e-30          4.522445e-30                1   \n",
      "75124          2.935075e-85          2.935075e-85                1   \n",
      "75125          5.157302e-93          5.157302e-93                1   \n",
      "\n",
      "       priority_Score  \n",
      "0                 100  \n",
      "1                 100  \n",
      "2                 100  \n",
      "3                 100  \n",
      "4                 100  \n",
      "...               ...  \n",
      "75121               0  \n",
      "75122               0  \n",
      "75123               0  \n",
      "75124               0  \n",
      "75125               0  \n",
      "\n",
      "[75126 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load MuPeXi results (Assuming it's a TSV file)\n",
    "\n",
    "df = pd.read_csv(\"C:/Users/dell/Downloads/WD22.mupexi.txt\", sep=\"\\t\", skiprows=5)\n",
    "\n",
    "\n",
    "# Apply filters for tumor neoantigens\n",
    "neoantigens = df[(df[\"Norm_MHCrank_EL\"] > 2) & (df[\"Mut_MHCrank_EL\"] < 2)]\n",
    "\n",
    "# Apply filters for strong tumor neoantigens\n",
    "strong_neoantigens = neoantigens[neoantigens[\"Mut_MHCrank_EL\"] < 0.5]\n",
    "\n",
    "# Count unique neoantigen peptides per tumor\n",
    "neoantigen_counts = neoantigens.groupby(\"Gene_Symbol\").size()\n",
    "\n",
    "# Count strong neoantigens per tumor\n",
    "strong_neoantigen_counts = strong_neoantigens.groupby(\"Gene_Symbol\").size()\n",
    "\n",
    "# Print results\n",
    "print(\"Total Tumor Neoantigens per Tumor Sample of WD22:\")\n",
    "print(neoantigen_counts)\n",
    "\n",
    "print(\"\\nTotal Strong Tumor Neoantigens per Tumor Sample of WD22:\")\n",
    "print(strong_neoantigen_counts)\n",
    "\n",
    "print(f\"Total Tumour Neoantigens of WD22: {len(neoantigens)}\")\n",
    "print(f\"Total Strong Tumour Neoantigens of WD22: {len(strong_neoantigens)}\")\n",
    "\n",
    "# Save results to Excel\n",
    "neoantigens.to_excel(\"Neoantigens_WD22.xlsx\", index=False)\n",
    "strong_neoantigens.to_excel(\"Strong_Neoantigens_WD22.xlsx\", index=False)\n",
    "\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9d6a456-14c4-4d8f-930e-787da9448672",
   "metadata": {},
   "source": [
    "LAQ 3-Part 1:\r\n",
    "(1) How many tumour neoantigens can you find for this tumour?4945\r\n",
    "(2) How many strong tumour neoantigens are there for this tumour70 88\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9acf9bd8-20e7-47ba-8dd2-80ac66178f48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     HLA_allele Neoantigen Peptide  Normal Rank Score  Tumor Rank Score  \\\n",
      "0    HLA-A24:02         .........S             7.2538            0.0156   \n",
      "1    HLA-A24:02          ........S             7.2642            0.0198   \n",
      "2    HLA-A24:02          ........S             7.5499            0.0316   \n",
      "3    HLA-A24:02          ........S             9.5501            0.0417   \n",
      "4    HLA-A02:01          ........G            11.6109            0.0767   \n",
      "..          ...                ...                ...               ...   \n",
      "176  HLA-A02:01          ...V.....             2.1710            0.1317   \n",
      "177  HLA-A24:02         .........L             2.1402            0.3818   \n",
      "178  HLA-A24:02          .R.......             2.1333            0.1381   \n",
      "179  HLA-A02:01          ..R......             2.0681            0.4954   \n",
      "180  HLA-A24:02        ..........L             2.0317            0.4550   \n",
      "\n",
      "    Affected Gene Tumor Sample  \n",
      "0        SIGLEC12         WD06  \n",
      "1        C1orf186         WD06  \n",
      "2          OR51B2         WD06  \n",
      "3           CNBD1         WD06  \n",
      "4            DRC1         WD06  \n",
      "..            ...          ...  \n",
      "176         OR5K3         WD22  \n",
      "177       TAS2R38         WD22  \n",
      "178         OR1S1         WD22  \n",
      "179         TACC2         WD22  \n",
      "180          TLX3         WD22  \n",
      "\n",
      "[181 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "df_06 = pd.read_excel(\"Strong_Neoantigens_WD6.xlsx\", usecols=[\"Gene_Symbol\", \"Norm_peptide\", \"Mut_MHCrank_EL\", \"Norm_MHCrank_EL\", \"HLA_allele\"])\n",
    "df_13 = pd.read_excel(\"Strong_Neoantigens_WD13.xlsx\", usecols=[\"Gene_Symbol\", \"Norm_peptide\", \"Mut_MHCrank_EL\", \"Norm_MHCrank_EL\", \"HLA_allele\"])\n",
    "df_14 = pd.read_excel(\"Strong_Neoantigens_WD14.xlsx\", usecols=[\"Gene_Symbol\", \"Norm_peptide\", \"Mut_MHCrank_EL\", \"Norm_MHCrank_EL\", \"HLA_allele\"])\n",
    "df_22 = pd.read_excel(\"Strong_Neoantigens_WD22.xlsx\", usecols=[\"Gene_Symbol\", \"Norm_peptide\", \"Mut_MHCrank_EL\", \"Norm_MHCrank_EL\", \"HLA_allele\"])\n",
    "\n",
    "df_06[\"Tumor Sample\"] = \"WD06\"\n",
    "df_13[\"Tumor Sample\"] = \"WD13\"\n",
    "df_14[\"Tumor Sample\"] = \"WD14\"\n",
    "df_22[\"Tumor Sample\"] = \"WD22\"\n",
    "\n",
    "concatenated_df = pd.concat([df_06, df_13, df_14, df_22], ignore_index=True)\n",
    "\n",
    "concatenated_df.columns = [\"HLA_allele\", \"Neoantigen Peptide\", \"Normal Rank Score\", \"Tumor Rank Score\", \"Affected Gene\", \"Tumor Sample\"]\n",
    "\n",
    "concatenated_df.to_csv(\"strong_tumor_neoantigens_all_samples.csv\", index=False)\n",
    "\n",
    "print(concatenated_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1118d7b9-bd8d-41b5-a71d-ab3558954595",
   "metadata": {},
   "source": [
    "LAQ 3- Part 3\r\n",
    "Output of this part of the question are stored in the LAQ_3 -> Output_fies3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "255bfb37-b1d9-430c-a222-3d4dfc647529",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Neoantigen Peptide\n",
      "..........C    1\n",
      "..........L    1\n",
      "..........P    1\n",
      "..........R    2\n",
      "..........S    2\n",
      "              ..\n",
      "P........      3\n",
      "SK.......E     1\n",
      "T......P.E     1\n",
      "VT......P      1\n",
      "VT......P.E    1\n",
      "Name: Tumor Sample, Length: 87, dtype: int64\n",
      "Neoantigen Peptide\n",
      "..........R    2\n",
      "..........S    2\n",
      ".........C     2\n",
      ".........L     2\n",
      ".........P     2\n",
      ".........R     2\n",
      ".........S     2\n",
      "........C      3\n",
      "........G      2\n",
      "........P      4\n",
      "........R      3\n",
      "........S      3\n",
      "........T      3\n",
      ".......S       2\n",
      "..R......      2\n",
      ".C.......      2\n",
      ".C........     2\n",
      ".D.......      2\n",
      ".H.......      4\n",
      ".P.......      3\n",
      ".R.......      3\n",
      ".R........     2\n",
      ".S.......      3\n",
      ".S........     2\n",
      "E........      2\n",
      "P........      3\n",
      "Name: Tumor Sample, dtype: int64\n",
      "     HLA_allele Neoantigen Peptide  Normal Rank Score  Tumor Rank Score  \\\n",
      "0    HLA-A24:02         .........S             7.2538            0.0156   \n",
      "1    HLA-A24:02          ........S             7.2642            0.0198   \n",
      "2    HLA-A24:02          ........S             7.5499            0.0316   \n",
      "3    HLA-A24:02          ........S             9.5501            0.0417   \n",
      "4    HLA-A02:01          ........G            11.6109            0.0767   \n",
      "..          ...                ...                ...               ...   \n",
      "171  HLA-A02:01          ........S             2.4797            0.0522   \n",
      "175  HLA-A24:02          .P.......             2.2925            0.4936   \n",
      "177  HLA-A24:02         .........L             2.1402            0.3818   \n",
      "178  HLA-A24:02          .R.......             2.1333            0.1381   \n",
      "179  HLA-A02:01          ..R......             2.0681            0.4954   \n",
      "\n",
      "    Affected Gene Tumor Sample  \n",
      "0        SIGLEC12         WD06  \n",
      "1        C1orf186         WD06  \n",
      "2          OR51B2         WD06  \n",
      "3           CNBD1         WD06  \n",
      "4            DRC1         WD06  \n",
      "..            ...          ...  \n",
      "171        PPP3CB         WD22  \n",
      "175        MYBPC2         WD22  \n",
      "177       TAS2R38         WD22  \n",
      "178         OR1S1         WD22  \n",
      "179         TACC2         WD22  \n",
      "\n",
      "[117 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Load the CSV data\n",
    "shared = pd.read_csv(\"strong_tumor_neoantigens_all_samples.csv\")\n",
    "\n",
    "\n",
    "# Group by the 'Neoantigen Peptide' and count how many unique tumor samples have each neoantigen\n",
    "shared_neoantigens = shared.groupby(\"Neoantigen Peptide\")[\"Tumor Sample\"].nunique()\n",
    "\n",
    "print(shared_neoantigens)\n",
    "\n",
    "# Filter for neoantigens that are present in more than one sample\n",
    "shared_neoantigens = shared_neoantigens[shared_neoantigens > 1]\n",
    "\n",
    "# Display the shared neoantigens\n",
    "print(shared_neoantigens)\n",
    "\n",
    "# Optional: You can also extract the full rows for the shared neoantigens\n",
    "shared_neoantigen_rows = shared[shared[\"Neoantigen Peptide\"].isin(shared_neoantigens.index)]\n",
    "print(shared_neoantigen_rows)\n",
    "\n",
    "# Save the shared neoantigens to a CSV (optional)\n",
    "shared_neoantigen_rows.to_csv(\"shared_tumor_neoantigens.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef171666-f0b6-49d5-b5c1-276386a0c349",
   "metadata": {},
   "source": [
    "Neoantigen-based therapies utilize tumor-specific antigens derived from cancer mutations to stimulate immune responses. \n",
    "\n",
    "1). Positive Implications\n",
    "1.1 Personalized Treatment:\n",
    "Tailored to an individual’s unique cancer mutations, enhancing efficacy while minimizing off-target effects.\n",
    "\n",
    "1.2 Reduced Side Effects:\n",
    "Since neoantigens are tumor-specific, they significantly lower the risk of autoimmune reactions and damage to healthy tissue.\n",
    "\n",
    "1.3 Synergy with Other Therapies:\n",
    "Can be combined with immunotherapies such as checkpoint inhibitors to enhance overall treatment effectiveness.\n",
    "\n",
    "2). Challenges and Limitations\n",
    "2.1 Complexity and Cost:\n",
    "Requires extensive genomic analysis, bioinformatics assessment, and personalized manufacturing, increasing both cost and complexity.\n",
    "\n",
    "2.2 Uncertain Immunogenicity:\n",
    "Not all neoantigens elicit strong immune responses, making reliable identification a challenge.\n",
    "\n",
    "2.3 Tumor Heterogeneity and Immune Evasion:\n",
    "Variability within tumors may result in incomplete targeting, allowing some cancer cells to escape immune detection and potentially leading to recurrence.\n",
    "\n",
    "2.4 Lengthy Development Process:\n",
    "Personalized production and validation can take weeks or months, limiting accessibility for patients with rapidly progressing cancers.\n",
    "\n",
    "2.5 Ethical and Accessibility Concerns:\n",
    "High costs and the need for specialized facilities may restrict availability, exacerbating healthcare disparities.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "2) Challenges and Limitations\n",
    "2.1 Complexity and Cost:\n",
    "Requires extensive genomic analysis, bioinformatics assessment, and personalized manufacturing, increasing both cost and complexity.\n",
    "\n",
    "2.2 Uncertain Immunogenicity:\n",
    "Not all neoantigens elicit strong immune responses, making reliable identification a challenge.\n",
    "\n",
    "2.3 Tumor Heterogeneity and Immune Evasion:\n",
    "Variability within tumors may result in incomplete targeting, allowing some cancer cells to escape immune detection and potentially leading to recurrence.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
